synapse/util/caches/descriptors.py

```diff
@@ -182,16 +188,28 @@ def __get__(self, obj, objtype=None):
             name=self.orig.__name__,
             max_entries=self.max_entries,
             keylen=self.num_args,
-            lru=self.lru,
             tree=self.tree,
         )
 
         @functools.wraps(self.orig)
         def wrapped(*args, **kwargs):
+            cache_context = kwargs.pop("cache_context", None)
```

Comment:
I would probably call this "on_invalidate" and have it take a callback rather than a "cache_context" object.